api/src/routes/ical.ts

```diff
@@ -0,0 +1,73 @@
+import { ICalCalendar } from "ical-generator";
+import { Request, Response, Handler } from "express";
+import { Pool } from "pg";
+import slugify from "slugify";
+
+type CtfRow = {
+  id: number;
+  title: string;
+  start_time: string;
+  end_time: string;
+  ctf_url: string;
+  description: string;
+};
+
+type IcalPasswordRow = {
+  ical_password: string;
+};
+
+export function icalRoute(pool: Pool): Handler {
+  async function checkIcalPassword(
+    userPass: string | undefined
+  ): Promise<boolean> {
+    const r = await pool.query<IcalPasswordRow>(
+      "SELECT ical_password FROM ctfnote.settings"
+    );
+    const db_password = r.rows[0].ical_password;
+    // If the password is null or empty allow any user
+    if (!db_password) return true;
+    return db_password === userPass;
+  }
+
+  async function getCtfs(): Promise<CtfRow[]> {
+    const r = await pool.query<CtfRow>(
+      "SELECT id, title, start_time, end_time, ctf_url, description FROM ctfnote.ctf"
+    );
+
+    return r.rows;
+  }
+
+  return async function (req: Request, res: Response): Promise<void> {
+    const { key } = req.query;
+
+    if (
+      !(typeof key == "string" || key == undefined) ||
+      !(await checkIcalPassword(key))
+    ) {
+      res.status(403);
+      res.send("Forbidden\n");
+      return;
+    }
+
+    const cal = new ICalCalendar();
+    const ctfs = await getCtfs();
+
+    for (const ctf of ctfs) {
+      // Assumes that CTFNote is hosted at the root of the domain
+      const ctf_url = new URL(
+        `/#/ctf/${ctf.id}-${slugify(ctf.title)}/info`,
+        `${req.protocol}://${req.headers.host}`
```

Comment:
Hmmm… this is a problem.
`req.headers.host` can contain all kind of nasty stuff.

For example :
```
curl 'http://127.0.0.1:8080/calendar.ics?key=5895ec74ae0decd41bf5b09417f1eef2' --header 'Host: @'
<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx</center>
</body>
</html>
```

This causes an uncaught exception :
```
tmphwaeptjw1g-api-1  | node:internal/errors:464
tmphwaeptjw1g-api-1  |     ErrorCaptureStackTrace(err);
tmphwaeptjw1g-api-1  |     ^
tmphwaeptjw1g-api-1  | 
tmphwaeptjw1g-api-1  | TypeError [ERR_INVALID_URL]: Invalid URL
tmphwaeptjw1g-api-1  |     at new NodeError (node:internal/errors:371:5)
tmphwaeptjw1g-api-1  |     at onParseError (node:internal/url:552:9)
tmphwaeptjw1g-api-1  |     at new URL (node:internal/url:628:5)
tmphwaeptjw1g-api-1  |     at new URL (node:internal/url:625:22)
tmphwaeptjw1g-api-1  |     at /app/dist/routes/ical.js:101:39
tmphwaeptjw1g-api-1  |     at step (/app/dist/routes/ical.js:33:23)
tmphwaeptjw1g-api-1  |     at Object.next (/app/dist/routes/ical.js:14:53)
tmphwaeptjw1g-api-1  |     at fulfilled (/app/dist/routes/ical.js:5:58)
tmphwaeptjw1g-api-1  |     at processTicksAndRejections (node:internal/process/task_queues:96:5) {
tmphwaeptjw1g-api-1  |   input: 'http://@',
tmphwaeptjw1g-api-1  |   code: 'ERR_INVALID_URL'
tmphwaeptjw1g-api-1  | }
tmphwaeptjw1g-api-1  | error Command failed with exit code 1.
tmphwaeptjw1g-api-1  | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
tmphwaeptjw1g-api-1 exited with code 1
```

This stops the `api` container, resulting in a denial of service.
I don't think it is possible to get a higher impact from this (prove me wrong ! :-) )

Reply:
Couldn't we just put this in a try-catch and just stop if the header contains errors? I think it is safe to assume that _normally_ the header would be correct, unless you modify it like a real hacker.

Reply:
Sure, I'll fix it that way.

In our fork we're already running into issues with this implementation though, because we're serving the calendar from a different reverse proxy than CTFNote, so the generated URL is not correct 😅 .

Reply:
@frereit, have you already fixed this in your fork? If so, is that fix similar to my suggested fix?

Reply:
Sorry for the delay, I've been busy and on holiday. No, no fix has been implemented on our side yet. I will do so in the coming week.

Reply:
Okay, I removed the crash on invalid host.

Note: I now "trust" the user that the X-Forwarded-Host header is set correctly, or not at all.
This isn't an issue though. If the user sets an invalid X-Forwarded-Host header it will simply be reflected in the ical, yielding either an invalid link or an entirely invalid ical. This is really just a self-own and not a problem imo, because "garbage request yields garbage data" is not a vulnerability per se.

Reply:
I totally agree with that. And the fix does indeed not crash the server anymore, so I think this is a good solution.